### DISCUSSION TITLE
When a global client is used for 2 connections (wrongly), the 2nd connection gets the 1st connection message.

### DISCUSSION THREAD

**Discussion by qiulang** [socketio/socket.io#4304] [Q&A] [answered]
When I try to fix a memory leak and also try to test `autoUnref` feature in 4.x I find that if a global client is used for 2 connections (wrongly), the 2nd connection gets the 1st connection message. Check the following code (also attached the full ts code)

Some explanation for the key points. The code declare `socket` as a global variable. This is wrong, because it will use for 2 different connections as `main()` shows.

My server logic is simple, when client connects it needs to send a login message with user_id , if the server finds the same user_id login with a different socket.id, it decides that this is the case that the same user login from a different client so it will emit `logoff`  to the first socket. When the socket gets `logoff` message it will close its connection.

The correct way is to make `socket` a function scope variable. But I find when it is a global variable, the second connection gets the logoff message. Why is that ?

[socketConnectBug.log](https://github.com/socketio/socket.io/files/8175518/socketConnectBug.log) Github doesn't allow to upload ts file so I change the file extension to log.

The server codes
```
io.on("connection", (socket) => {
  socket.on("login",({user_id,client},fn) =>{
      log(`${user_id}:${client} using ${socket.id} logged in`)
      fn("accept")
      let old_socket = records[user_id]
      if ( old_socket ) {
        log(`The same ${user_id} with ${old_socket} has logged in, let him log off first`)
        io.to(old_socket).emit('logoff', 'another login')
      }
      records[user_id] = socket.id
  })
  socket.on("disconnect",(reason) => {
      log(`${socket.id} disconnected with reason ${reason}`)
  })
});
```
The client code 

```
const io = require('socket.io-client')
const server = 'http://localhost:3000'
const user_id = '1_2_1000'
let socket //wrong!! Should NOT use a global variable
let wsBit = 0

function setupWS(client) {
//   let socket = io(server, { transports: ['websocket']}) //correct, should use a function scope variable
  socket = io(server, { transports: ['websocket']}) //wrong!! Should NOT use a global variable
  socket.on('connect', function () {
    log(`${client} connected to local ws server with ${socket.id}`)
    socket.emit('login', { user_id, client }, reply => {
        wsBit +=1
        log(`Current connection:${wsBit}`) 
    })
  })
  socket.on('logoff', (msg) => {
    log(`${socket.id}:${client} should be kicked off for: ${msg}`)
    socket.close()
  })
  socket.on('disconnect',(msg) =>{
    log(`${socket.id}: disconnect: ${msg}`)
    wsBit -=1
    log(`Current connection:${wsBit}`) 
  })
}

function start(client) {
  setupWS(client)
  log(`Start to connect to ws server:${server} ... using client:${client}`)
}

function main() {
  start('nodejs')
  setTimeout(()=> {
    log('open another connection in 5 seconds')
    start("browser")
  },5000)
}

main()
```

The result shows the second connection gets the logoff message. From the server logs it also confirms it is the 2nd connection disconnected. Why is that ?

```
Start to connect to ws server:http://localhost:3000 ... using client:nodejs
Connected to local ws server with 5sxucsyNrJqzB0LEAAAB
ws connected with nodejs
open another connection in 5 seconds
Start to connect to ws server:http://localhost:3000 ... using client:browser
Connected to local ws server with qA7A9O8EiyxdtJbuAAAD
qA7A9O8EiyxdtJbuAAAD:nodejs should be kicked off for: another login
undefined: disconnect: io client disconnect
Failed to connect local ws server with browser
```

**Comment by darrachequesne** [comment, marked as answer]
Hi! There is a lot of global variables in your example! I'm not sure to completely understand your use case though.

Could you please reduce it to the bare minimum, and include the code for the server? Thanks :+1:

**Reply by qiulang** [reply]
Sorry for the late reply. I have been having a bad cold and crushed for 7 days. I will try to reduce the server code to the bare minimum and include it here. But I think I already reduce the client code to the bare minimum.

**Reply by qiulang** [reply]
@darrachequesne I have updated my question, reduce the code to the bare minimum and also paste the server code.

I also create a simple repo to demonstrate it https://github.com/qiulang/2sockets

Can you take a look ? Thanks!

**Reply by qiulang** [reply]
I read the socket.io-client code but I still can't find out why the 2nd connection gets the 1st connection message. Now I can only resort to these words on stackoverfolw, which in themselves are easy to understand, [Why shouldn't I use global variables in JavaScript for something that's constant?](https://stackoverflow.com/questions/6989903/why-shouldnt-i-use-global-variables-in-javascript-for-something-thats-constant)

> If you have any asynchronous code that modifies globals or timer-driven code that modifies globals and more than one asynchronous operation can be in flight at the same time, the multiple async operations can step on each other through the modification of the same globals.

**Reply by qiulang** [reply]
I finally understand that it is my log that gave me the wrong impression that the 2nd connection gets the 1st connection message. 
So the lesson I learn is that I may introduce global variable unintentionally especially when not using strict mode 🥵

I got my answer from this SO discussion [Global variables in asynchronous code (socket.io) show an error in an unexpected way](https://stackoverflow.com/questions/71852823/global-variables-in-asynchronous-code-socket-io-show-an-error-in-an-unexpected)
